syntax = "proto3";

package contextbrain;

// =====================================================
// Core Brain Service - Universal Knowledge Operations
// =====================================================

enum SourceType {
    SOURCE_TYPE_UNSPECIFIED = 0;
    SOURCE_TYPE_CHUNK = 1;
    SOURCE_TYPE_CONCEPT = 2;
}

// =====================================================
// Search Messages
// =====================================================

message SearchRequest {
    string tenant_id = 1;
    string query_text = 2;
    repeated float query_vector = 3;
    int32 limit = 4;
    float min_score = 5;
    repeated string source_types = 6;
}

message SearchResult {
    string id = 1;
    string content = 2;
    float score = 3;
    string source_type = 4;
    map<string, string> metadata = 5;
}

message SearchResponse {
    repeated SearchResult results = 1;
}

// =====================================================
// Knowledge Graph Messages
// =====================================================

message GraphSearchRequest {
    string tenant_id = 1;
    repeated string entrypoint_ids = 2;
    int32 max_hops = 3;
    repeated string allowed_relations = 4;
}

message GraphEdge {
    string source_id = 1;
    string target_id = 2;
    string relation = 3;
    float weight = 4;
}

message GraphSearchResponse {
    repeated SearchResult nodes = 1;
    repeated GraphEdge edges = 2;
}

message CreateKGRelationRequest {
    string tenant_id = 1;
    string source_type = 2;
    string source_id = 3;
    string relation = 4;
    string target_type = 5;
    string target_id = 6;
}

message CreateKGRelationResponse {
    bool success = 1;
}

// =====================================================
// Upsert Messages
// =====================================================

message UpsertRequest {
    string tenant_id = 1;
    string content = 2;
    string source_type = 3;
    map<string, string> metadata = 4;
}

message UpsertResponse {
    string id = 1;
    bool success = 2;
}

// =====================================================
// NewsEngine Messages (Pink Pony)
// =====================================================

message NewsItem {
    string id = 1;
    string tenant_id = 2;
    string url = 3;
    string headline = 4;
    string summary = 5;
    string category = 6;
    string source_api = 7;  // "perplexity" or "serper"
    map<string, string> metadata = 8;
    string harvested_at = 9;
}

message UpsertNewsItemRequest {
    string tenant_id = 1;
    NewsItem item = 2;
    string item_type = 3;  // "raw" or "fact"
}

message UpsertNewsItemResponse {
    string id = 1;
    bool success = 2;
    string message = 3;
}

message GetNewsItemsRequest {
    string tenant_id = 1;
    string item_type = 2;  // "raw" or "fact"
    int32 limit = 3;
    string since = 4;  // ISO datetime
}

message GetNewsItemsResponse {
    repeated NewsItem items = 1;
}

message NewsPost {
    string id = 1;
    string tenant_id = 2;
    string fact_id = 3;
    string agent = 4;
    string headline = 5;
    string content = 6;
    string emoji = 7;
    string fact_url = 8;
    string scheduled_at = 9;
    string published_at = 10;
}

message UpsertNewsPostRequest {
    string tenant_id = 1;
    NewsPost post = 2;
}

message UpsertNewsPostResponse {
    string id = 1;
    bool success = 2;
}

// =====================================================
// Core Brain Service Definition
// =====================================================

service BrainService {
    // Semantic/Hybrid search
    rpc Search(SearchRequest) returns (SearchResponse);

    // Graph Traversal
    rpc GraphSearch(GraphSearchRequest) returns (GraphSearchResponse);

    // Create Knowledge Graph relation
    rpc CreateKGRelation(CreateKGRelationRequest) returns (CreateKGRelationResponse);
    
    // Upsert content
    rpc Upsert(UpsertRequest) returns (UpsertResponse);
    
    // NewsEngine operations
    rpc UpsertNewsItem(UpsertNewsItemRequest) returns (UpsertNewsItemResponse);
    rpc GetNewsItems(GetNewsItemsRequest) returns (GetNewsItemsResponse);
    rpc UpsertNewsPost(UpsertNewsPostRequest) returns (UpsertNewsPostResponse);
}

